x/ccv/provider/keeper/punish_validator.go

```diff
@@ -31,8 +29,68 @@ func (k Keeper) JailValidator(ctx sdk.Context, providerAddr types.ProviderConsAd
 		k.stakingKeeper.Jail(ctx, providerAddr.ToSdkConsAddr())
 	}
 
-	// update jail time to end after double sign jail duration
+	// Jail the validator to trigger the unbonding of the validator
+	// (see cosmos/cosmos-sdk/blob/v0.47.0/x/staking/keeper/val_state_change.go#L194).
 	k.slashingKeeper.JailUntil(ctx, providerAddr.ToSdkConsAddr(), evidencetypes.DoubleSignJailEndTime)
 
-	// TODO: add tombstoning back once we integrate the slashing
+	// Tombstone the validator so that we cannot slash the validator more than once
+	// (see cosmos/cosmos-sdk/blob/v0.47.0/x/evidence/keeper/infraction.go#L94).
+	// Note that we cannot simply use the fact that a validator is jailed to avoid slashing more than once
+	// because then a validator could i) perform an equivocation, ii) get jailed (e.g., through downtime)
+	// and in such a case the validator would not get slashed when calling `SlashValidator`.
+	// TODO: check if tombstone ... can it panic???
+	k.slashingKeeper.Tombstone(ctx, providerAddr.ToSdkConsAddr())
+}
+
+func (k Keeper) ComputePowerToSlash(undelegations []stakingtypes.UnbondingDelegation,
+	redelegations []stakingtypes.Redelegation, power int64, powerReduction sdk.Int) int64 {
+
+	// compute the total numbers of tokens currently being undelegated
+	undelegationsInTokens := sdk.NewInt(0)
+	for _, u := range undelegations {
+		for _, entry := range u.Entries {
```

Comment:
Must exclude entries that reached maturity, see https://github.com/cosmos/cosmos-sdk/blob/0af2f4da004cbea6414a8bad56e8bdcd45badf1e/x/staking/keeper/slash.go#L200. Maybe it would be better to just use the `SlashUnbondingDelegation` method from the staking module, but calling it from a cached context.

Reply:
Good catch. Fixed it. Current approach uses the same condition to skip entries as the one in [Slash](https://github.com/cosmos/cosmos-sdk/blob/0af2f4da004cbea6414a8bad56e8bdcd45badf1e/x/staking/keeper/slash.go#L200).

I believe that calling `SlashUnbondingDelegation` directly might complicate things because when we call the staking's module `Slash` method,  `Slash` calls `SlashUnbondingDelegation` and `SlashRedelegation` so we would need to figure out a way to avoid slashing twice the same undelegations and redelegations.